pkg/resource/deploy/step.go

```diff
@@ -1693,3 +1696,69 @@ func getProvider(s Step, override plugin.Provider) (plugin.Provider, error) {
 	}
 	return provider, nil
 }
+
+// DiffStep isn't really a step like a normal step. It's just a way to get access to the parallel but bounded step
+// workers. We use this step to call `provider.Diff` in parallel with other steps.
+type DiffStep struct {
+	deployment    *Deployment                                  // the deployment that produced this diff
+	pcs           *promise.CompletionSource[plugin.DiffResult] // the completion source for this diff
+	old           *resource.State                              // the old resource state
+	new           *resource.State                              // the new resource state
+	ignoreChanges []string                                     // a list of property paths to ignore when diffing
+}
+
+func NewDiffStep(
+	deployment *Deployment, pcs *promise.CompletionSource[plugin.DiffResult], old, new *resource.State,
+	ignoreChanges []string,
+) Step {
+	return &DiffStep{
+		deployment:    deployment,
+		pcs:           pcs,
+		old:           old,
+		new:           new,
+		ignoreChanges: ignoreChanges,
+	}
+}
+
+func (s *DiffStep) Op() display.StepOp {
+	return OpDiff
+}
+
+func (s *DiffStep) Deployment() *Deployment { return s.deployment }
+func (s *DiffStep) Type() tokens.Type       { return s.new.Type }
+func (s *DiffStep) Provider() string        { return s.new.Provider }
+func (s *DiffStep) URN() resource.URN       { return s.new.URN }
+func (s *DiffStep) Old() *resource.State    { return s.old }
+func (s *DiffStep) New() *resource.State    { return s.new }
+func (s *DiffStep) Res() *resource.State    { return s.new }
+func (s *DiffStep) Logical() bool           { return true }
+
+func (s *DiffStep) Apply() (resource.Status, StepCompleteFunc, error) {
+	// DiffStep is a special step in that we're just using it as a way to get access to the parallel step
+	// workers. We don't actually want it to particapte in the rest of what normally happens for step
+	// execution. As such we never actually return an error here, we just reject the completion source.
```

Comment:
I assume we mean that we only reject in error cases? Also, what happens if we reject then, given that no error occurs? Will the continue logic turn that into a "real" stepgen error later on? Might be worth clarifying that here.